File: Anime Game Remap (for all users)/api/src/cpp/core/src/model/strategies/texEditors/texFilters/TransparencyAdjustFilter.cpp
// ##### Credits

// ===== Anime Game Remap (AG Remap) =====
// Authors: Albert Gold#2696, NK#1321
//
// if you used it to remap your mods pls give credit for "Albert Gold#2696" and "Nhok0169"
// Special Thanks:
//   nguen#2011 (for support)
//   SilentNightSound#7430 (for internal knowdege so wrote the blendCorrection code)
//   HazrateGolabi#1364 (for being awesome, and improving the code)

// ##### EndCredits

#include "AGRemapCore/model/strategies/texEditors/texFilters/TransparencyAdjustFilter.h"

#include <utility>

#include "AGRemapCore/model/files/TextureFile.h"
#include "AGRemapCore/model/strategies/texEditors/pixelTransforms/Transparency.h"

namespace AGRemapCore {

    TransparencyAdjustFilter::TransparencyAdjustFilter(int alphaChange, std::optional<ColourOrRangeSet> coloursToFilter):
        alphaChange(alphaChange), coloursToFilter(std::move(coloursToFilter)) {}

    void TransparencyAdjustFilter::adjustTransparency(TextureFile &texFile) {
        Transparency pixelTransform(alphaChange);
        int width = texFile.getWidth();
        int height = texFile.getHeight();

        for (int y = 0; y < height; ++y) {
            for (int x = 0; x < width; ++x) {
                Colour pixel = texFile.getPixel(x, y);
                pixelTransform.transform(pixel, x, y);
                texFile.setPixel(x, y, pixel);
            }
        }
    }

    void TransparencyAdjustFilter::transform(TextureFile &texFile) {
        if (!coloursToFilter.has_value()) {
            adjustTransparency(texFile);
            return;
        }

        Transparency pixelTransform(alphaChange);
        int width = texFile.getWidth();
        int height = texFile.getHeight();

        for (int y = 0; y < height; ++y) {
            for (int x = 0; x < width; ++x) {
                Colour pixel = texFile.getPixel(x, y);

                bool shouldAdjust = false;
                for (const auto &colourOrRange : *coloursToFilter) {
                    if (matchesColourOrRange(colourOrRange, pixel)) {
                        shouldAdjust = true;
                        break;
                    }
                }

                if (shouldAdjust) {
                    pixelTransform.transform(pixel, x, y);
                    texFile.setPixel(x, y, pixel);
                }
            }
        }
    }
}

File: Anime Game Remap (for all users)/api/src/py/FixRaidenBoss2/model/strategies/texEditors/texFilters/TransparencyAdjustFilter.py
##### Credits

# ===== Anime Game Remap (AG Remap) =====
# Authors: Albert Gold#2696, NK#1321
#
# if you used it to remap your mods pls give credit for "Albert Gold#2696" and "Nhok0169"
# Special Thanks:
#   nguen#2011 (for support)
#   SilentNightSound#7430 (for internal knowdege so wrote the blendCorrection code)
#   HazrateGolabi#1364 (for being awesome, and improving the code)

##### EndCredits

##### CppLocalImports
from .....core import CppTransparencyAdjustFilter
##### EndCppLocalImports


##### Script
class TransparencyAdjustFilter(CppTransparencyAdjustFilter):
    """
    This class inherits from :class:`CppTransparencyAdjustFilter`

    Adjust the trasparency (alpha channel) for an image

    :raw-html:`<br />`

    .. container:: operations

        **Supported Operations:**

        .. describe:: x(texFile)

            Calls :meth:`CppBaseTexFilter.transform` for the filter, ``x``

    Parameters
    ----------
    alphaChange: :class:`int`
        How much to adjust the alpha channel of each pixel. Range from -255 to 255

        .. note::
            The alpha channel for an image is inclusively bounded from 0 to 255

    coloursToFilter: Optional[Set[Union[:class:`Colour`, :class:`ColourRange`]]]
        The specific colours to have their transparency adjusted. If this value is ``None``, then will adjust the transparency for the entire image`<br />` :raw-html:`<br />`

        **Default**: ``None``
    """

    pass
##### EndScript
